### PR TITLE
WT-12049 Create test between incremental backup and background compaction

### PR DESCRIPTION
**Pull request by jiechenbo**
This ticket creates a python test to test the interaction between incremental backup and background compaction.

### CODE REVIEW

File: test/suite/test_compact11.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        # Compare the backups.
+        for uri in uris:
+            self.compare_backups(uri, self.backup_incr, self.backup_full)
+
+        # Background compaction may have been inspecting a table when disabled, which is considered
+        # as an interruption, ignore that message.
+        self.ignoreStdoutPatternIfExists('background compact interrupted by application')
+
+if __name__ == '__main__':
+    wttest.run()
```
Comment (etienneptl) [inline test/suite/test_compact11.py:197]:
```suggestion
    wttest.run()

```

File: test/suite/test_compact11.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        # Enable background compaction.
+        self.turn_on_bg_compact('free_space_target=1MB')
+
+        # Every time compaction makes progress, perform an incremental backup.
+        bytes_recovered = 0
+        while self.get_files_compacted(uris) < self.num_tables:
+            new_bytes_recovered = self.get_bytes_recovered()
+            if new_bytes_recovered != bytes_recovered:
+                self.prout('Compaction has made progress... ')
+                # Update the incremental backup ID from the parent class.
```
Comment (etienneptl) [inline test/suite/test_compact11.py:169]:
```suggestion
```

File: test/suite/test_compact11.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                bytes_recovered = new_bytes_recovered
+               
+        assert self.get_files_compacted(uris) == self.num_tables
+        self.turn_off_bg_compact()
```
Comment (etienneptl) [inline test/suite/test_compact11.py:173]: Do we need to turn it off?

File: test/suite/test_compact11.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+        # Write to disk.
+        self.session.checkpoint()
+        for uri in files:
+            bitmaps.append(self.parse_blkmods(uri))
+
+        # Enable background compaction.
+        self.turn_on_bg_compact('free_space_target=1MB')
+
+        # Every time compaction makes progress, perform an incremental backup.
```
Comment (etienneptl) [inline test/suite/test_compact11.py:163]: We are not doing this anymore at all?

File: test/suite/test_compact11.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            if new_bytes_recovered != bytes_recovered:
+                self.prout('Compaction has made progress... ')
+                # Update the incremental backup ID from the parent class.
+                bytes_recovered = new_bytes_recovered
+               
+        assert self.get_files_compacted(uris) == self.num_tables
+        self.turn_off_bg_compact()
+        self.prout(f'Compaction has compacted {self.num_tables} tables.')
+
+        self.bkup_id += 1
```
Comment (etienneptl) [inline test/suite/test_compact11.py:176]: Can you explain why you take an incremental backup here now?

File: test/suite/test_compact11.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import os, re, sys, time, wttest
```
Comment (etienneptl) [inline test/suite/test_compact11.py:29]:
```suggestion
import os, re, time, wttest
```

File: test/suite/test_compact11.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+from wiredtiger import stat
+from wtbackup import backup_base
+
+# test_compact11.py
+# Verify compaction XXX
```
Comment (etienneptl) [inline test/suite/test_compact11.py:34]: Please update the test summary

File: test/suite/test_compact11.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+            self.session.create(uri, self.create_params)
+
+        # Populate the first half of each table.
+        for uri in uris:
+            self.populate(uri, 0, self.table_numkv // 2, self.value_size)
+
+        # Write to disk.
+        self.session.checkpoint()
+
+        os.mkdir(self.backup_incr)
```
Comment (etienneptl) [inline test/suite/test_compact11.py:139]:
```suggestion
        # Take a full backup that will be used for incremental backups later during the test.
        os.mkdir(self.backup_incr)
```

File: test/suite/test_compact11.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        # Write to disk.
+        self.session.checkpoint()
+
+        # Delete 50% of each file.
+        for uri in uris:
+            self.delete_range(uri, 50 * self.table_numkv // 100)
+
+        # Write to disk.
+        self.session.checkpoint()
+        for uri in files:
```
Comment (etienneptl) [inline test/suite/test_compact11.py:157]:
```suggestion

        # Generate the bitmaps of each file.
        for uri in files:
```

Comment (jiechenbo) [reply test/suite/test_compact11.py:156]: [5138fc6](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/5138fc6cc4875fa9b291fc03b618948222bb21df)

File: test/suite/test_compact11.py (hunk 8 above)
Comment (jiechenbo) [reply test/suite/test_compact11.py:136]: [5138fc6](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/5138fc6cc4875fa9b291fc03b618948222bb21df)

File: test/suite/test_compact11.py (hunk 7 above)
Comment (jiechenbo) [reply test/suite/test_compact11.py]: [5138fc6](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/5138fc6cc4875fa9b291fc03b618948222bb21df)

File: test/suite/test_compact11.py (hunk 6 above)
Comment (jiechenbo) [reply test/suite/test_compact11.py]: [5138fc6](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/5138fc6cc4875fa9b291fc03b618948222bb21df)

File: test/suite/test_compact11.py (hunk 3 above)
Comment (jiechenbo) [reply test/suite/test_compact11.py]: I have removed it [5138fc6](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/5138fc6cc4875fa9b291fc03b618948222bb21df)

File: test/suite/test_compact11.py (hunk 2 above)
Comment (jiechenbo) [reply test/suite/test_compact11.py]: [5138fc6](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/5138fc6cc4875fa9b291fc03b618948222bb21df)

File: test/suite/test_compact11.py (hunk 1 above)
Comment (jiechenbo) [reply test/suite/test_compact11.py]: Update the incremental backup ID from the parent class.

File: test/suite/test_compact11.py (hunk 4 above)
Comment (jiechenbo) [reply test/suite/test_compact11.py]: Yes, there are a few reasons for the change. 
1.  How I understood the reason for the testing is to specifically test for how the incremental backup bitmap field is interacted with background compaction. If that is the case, we don't need multiple incremental backups to check this and only one.
2. Once a second incremental backup happens {{blk_parsemods}} can't be used to check if the bitmapping is correct and unfortunately the ordering of how we can read the {{blocks}} field is not ordered.

Comment (etienneptl) [reply test/suite/test_compact11.py]: I think having background compaction and incr backups in parallel is also an interesting scenario. If we try to keep this, could we compare the "just taken" incr backups straight away instead of waiting for the loop to finish? This way, we will not compare the nth iteration of an incr backup but the fresh new one.

Comment (jiechenbo) [reply test/suite/test_compact11.py]: done! I have made those changes in compact10 [5138fc6](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/5138fc6cc4875fa9b291fc03b618948222bb21df)

File: test/suite/test_compact11.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+        self.session.checkpoint()
+
+        # Delete 50% of each file.
+        for uri in uris:
+            self.delete_range(uri, 50 * self.table_numkv // 100)
+
+        # Write to disk.
+        self.session.checkpoint()
+
+        # Generate the bitmaps of each file.
```
Comment (sueloverso) [inline test/suite/test_compact11.py:155]: I suggest "Get the bitmaps ...". This code is only reading the bitmaps, not generating them.

File: test/suite/test_compact11.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+        for uri in files:
+            bitmaps.append(self.parse_blkmods(uri))
+
+        # Turn on background compaction to allow the bitmap blocks to be modified
+        # from compact operation.
+        self.turn_on_bg_compact('free_space_target=1MB')
+        while self.get_files_compacted(uris) < self.num_tables:
```
Comment (sueloverso) [inline test/suite/test_compact11.py:162]: I don't know how many times we loop here, but maybe take an incremental backup here so that we get multiple backups loops through here and can get multiple iterations.

**Review by sueloverso — COMMENTED** [review]
Generally this looks good. I have a question or two before merging.

File: test/suite/test_compact10.py [hunk 12]
```diff
@@ -140,12 +140,12 @@ def test_compact10_full_backup(self):
         assert self.get_files_compacted(uris) == self.num_tables
 
         # Take a second full backup.
-        os.mkdir(self.backup_dir_2)
-        self.take_full_backup(self.backup_dir_2)
+        os.mkdir(self.backup_incr)
```
Comment (etienneptl) [inline test/suite/test_compact10.py:143]: I am not a fan of this change as the name is misleading, we are not doing an incremental backup as the name may suggest.

File: test/suite/test_compact10.py [hunk 13]
```diff
@@ -140,12 +140,12 @@ def test_compact10_full_backup(self):
...
+        self.take_full_backup(self.backup_incr)
 
         # Compare the backups.
         for uri in uris:
-            self.compare_backups(uri, self.backup_dir_1, self.backup_dir_2)
+            self.compare_backups(uri, self.backup_full, self.backup_incr)
```
Comment (etienneptl) [inline test/suite/test_compact10.py:148]: Same here, we are comparing two full backups, not a full and an incremental one.

File: test/suite/test_compact10.py [hunk 14]
```diff
@@ -168,13 +168,12 @@ def test_compact10_incr_backup(self):
...
-        os.mkdir(self.backup_dir_1)
-        self.take_full_backup(self.backup_dir_1)
+        self.setup_directories(self.backup_incr, self.backup_full)
+        self.take_full_backup(self.backup_full)
 
         # - The second one is used for the incremental backups.
-        os.mkdir(self.backup_dir_2)
         self.initial_backup = True
-        self.take_full_backup(self.backup_dir_2)
+        self.take_full_backup(self.backup_incr)
```
Comment (etienneptl) [inline test/suite/test_compact10.py:176]: The naming here makes more sense. We could have dedicated variables for each test.

File: test/suite/test_compact10.py [hunk 15]
```diff
@@ -183,19 +182,25 @@ def test_compact10_incr_backup(self):
...
-                self.take_incr_backup(self.backup_dir_2)
-                bytes_recovered = new_bytes_recovered
-
-        assert self.get_files_compacted(uris) == self.num_tables
 
-        # Compare the backups.
-        for uri in uris:
-            self.compare_backups(uri, self.backup_dir_1, self.backup_dir_2)
+                self.take_incr_backup(self.backup_incr)
+                self.take_full_backup(self.backup_full)
```
Comment (etienneptl) [inline test/suite/test_compact10.py:193]: Why are we taking a full backup? As described at the start of the test, we want to perform only incremental backups as compaction is running.

File: test/suite/test_compact10.py [hunk 16]
```diff
@@ -183,19 +182,25 @@ def test_compact10_incr_backup(self):
...
+
+                # Compare backups function does a connection reopen, which will reset WiredTiger
```
Comment (etienneptl) [inline test/suite/test_compact10.py:195]: Where does it do a connection reopen in [compare_backups](https://github.com/wiredtiger/wiredtiger/blob/5be00f5be353e30d1b8c4155a1f32653770a6f9b/test/suite/wtbackup.py#L218)?

File: test/suite/test_compact10.py [hunk 17]
```diff
@@ -183,19 +182,25 @@ def test_compact10_incr_backup(self):
...
+                # statistics and background compaction. Keep record of how many files we have
```
[the remark this answers is not in the document]
Comment (jiechenbo) [reply test/suite/test_compact10.py]: I have rewrritten the logic of the python test [4cdad97](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/4cdad97f9da9cc86715e95394d6b77009a7b2343)

File: test/suite/test_compact10.py (hunk 16 above)
Comment (jiechenbo) [reply test/suite/test_compact10.py]: removed [4cdad97](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/4cdad97f9da9cc86715e95394d6b77009a7b2343)

File: test/suite/test_compact10.py (hunk 15 above)
Comment (jiechenbo) [reply test/suite/test_compact10.py]: fixed up now! [4cdad97](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/4cdad97f9da9cc86715e95394d6b77009a7b2343)

File: test/suite/test_compact10.py (hunk 14 above)
Comment (jiechenbo) [reply test/suite/test_compact10.py]: I have changed the naming [3570009](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/3570009210cdd4c711eea7aae6427731f792f903)

File: test/suite/test_compact10.py (hunk 13 above)
Comment (jiechenbo) [reply test/suite/test_compact10.py]: I have changed the naming [3570009](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/3570009210cdd4c711eea7aae6427731f792f903)

File: test/suite/test_compact10.py (hunk 12 above)
Comment (jiechenbo) [reply test/suite/test_compact10.py]: I have changed the naming [3570009](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/3570009210cdd4c711eea7aae6427731f792f903)

File: test/suite/test_compact10.py [hunk 18]
```diff
@@ -156,25 +156,27 @@ def test_compact10_full_backup(self):
...
+        os.makedirs(backup_full)
+        self.take_full_backup(backup_full)
 
-        # - The second one is used for the incremental backups.
-        os.mkdir(self.backup_dir_2)
+        # - The second one is used for the incremental backups and use a temporary backup
+        # directory to perform multiple incremental backups on different directories.
         self.initial_backup = True
-        self.take_full_backup(self.backup_dir_2)
+        os.makedirs(backup_incr)
```
Comment (etienneptl) [inline test/suite/test_compact10.py]:
```suggestion
        os.mkdir(backup_incr)
```

File: test/suite/test_compact11.py (hunk 11 above)
Comment (jiechenbo) [reply test/suite/test_compact11.py:162]: Unfortunately we can't take multiple incremental backups when we are test for the bitmap. The reason is because each incremental backup's bitmap will only show modified relative to the previous bitmap. What that means is that the compare_bitmaps will between two incremental bitmaps will not work, and there is no way of testing it. Let me provide an example to explain this better. E.g.
```
Base backup bitmap = 0000
1st Incr backup bitmap = 0011
2nd incr backup bitmap = 1100
```
If we see here, 2nd incremental bitmap shows that the first two bits are modified **compared to** the first incremental backup. There is nothing that we can test between the 2nd incremental and 1st incremental.

File: test/suite/test_compact10.py [hunk 19]
```diff
@@ -156,25 +156,27 @@ def test_compact10_full_backup(self):
...
+        self.take_full_backup(backup_incr)
+        shutil.copytree(backup_incr, self.home_tmp)
```
Comment (etienneptl) [inline test/suite/test_compact10.py:179]: Can you add a comment why we are using `copytree` here?

File: test/suite/test_compact11.py [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import os, re, time, wttest
+from wiredtiger import stat
+from wtbackup import backup_base
+
+# test_compact11.py
+# Verify background compaction and incremental backup behaviour. The block modifications bits in an
```
Comment (etienneptl) [inline test/suite/test_compact11.py:34]: This comment needs to be updated to better reflect the intent of the test.

File: test/suite/test_compact11.py [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+        # Update the incremental backup ID from the parent class.
+        self.bkup_id += 1
+        self.take_incr_backup(self.backup_incr)
+
+        self.pr('Comparing the original bitmaps with the new ones...')
+        for index, uri in enumerate(files):
+            new_bitmap = self.parse_blkmods(uri)
+            self.compare_bitmap(bitmaps[index], new_bitmap)
+
+        # Take a second full backup.
```
Comment (etienneptl) [inline test/suite/test_compact11.py:177]: This extra backup is not needed, we can remove it.

File: test/suite/test_compact10.py (hunk 19 above)
Comment (jiechenbo) [reply test/suite/test_compact10.py:179]: I have in the comment above it 
```
        # - The second one is used for the incremental backups and use a temporary backup
        # directory to perform multiple incremental backups on different directories.
```

File: test/suite/test_compact11.py (hunk 21 above)
Comment (jiechenbo) [reply test/suite/test_compact11.py]: Actually we need this backup because we should be using it to compare against the incremental backup. We expect no difference in data. I have moved this to earlier in the test before compaction starts [cc01737](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/cc01737af49a21ded94fdde966dff35ec018d323)

File: test/suite/test_compact11.py (hunk 20 above)
Comment (jiechenbo) [reply test/suite/test_compact11.py]: Updated [cc01737](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/cc01737af49a21ded94fdde966dff35ec018d323)

**Review by etienneptl — APPROVED** [review]
LGTM, thank you so much for taking the time to explain how backups are workng.

**Comment by etienneptl** [comment]
To fix the current failures, please merge with `develop`, `clang-analyzer` was removed recently.

File: test/suite/test_compact11.py [hunk 22]
```diff
@@ -31,8 +31,8 @@
 from wtbackup import backup_base
 
 # test_compact11.py
-# Verify background compaction and incremental backup behaviour. The block modifications bits in an
-# incremental backup should never be cleared.
+# Verify background compaction and backup behaviour. The block modifications bits in an
+# backup should never be cleared when background compact is working on tables.
```
Comment (sueloverso) [inline test/suite/test_compact11.py:35]: Why did you remove "incremental" here at the beginning of the line? The block mod bits are specifically only for incremental backup. (And it leaves the comment with '...in an backup' which is wrong grammatically.)

File: test/suite/test_compact11.py (hunk 11 above)
Comment (sueloverso) [reply test/suite/test_compact11.py:166]: It could work if each incremental used the original/first backup ID as the source instead of moving the source forward. For example, if the first full backup enabling incremental used "ID1", the first incremental could use "src=ID1,this=ID2" then the next "src=ID1,this=ID3", ...

I don't know if the python functions are set up this way, but it should work from WiredTiger's perspective and fix the bitmap comparison issue.

**Review by sueloverso — COMMENTED** [review]
I'll take another look if you think you can do the comparison in the loop

File: test/suite/wtbackup.py [hunk 23]
```diff
@@ -343,12 +343,14 @@ def take_log_backup(self, bkup_c, backup_dir, orig_logs, log_cursor=None):
     # Optional arguments:
     #   consolidate: Add consolidate option to the cursor.
```
Comment (etienneptl) [inline test/suite/wtbackup.py:344]: You may want to update the header describing the new args

File: test/suite/wtbackup.py [hunk 24]
```diff
@@ -343,12 +343,14 @@ def take_log_backup(self, bkup_c, backup_dir, orig_logs, log_cursor=None):
...
-    def take_incr_backup(self, backup_incr_dir, id=0, consolidate=False):
-        self.assertTrue(id > 0 or self.bkup_id > 0)
-        if id == 0:
+    def take_incr_backup(self, backup_incr_dir, src_id=0, dest_id=0, consolidate=False):
+        self.assertTrue(dest_id > 0 or self.bkup_id > 0)
+        if src_id == 0 and dest_id == 0:
             id = self.bkup_id
-        # Open the backup data source for incremental backup.
-        config = 'incremental=(src_id="ID' +  str(id - 1) + '",this_id="ID' + str(id) + '"'
+            src_id = id - 1
```
Comment (etienneptl) [inline test/suite/wtbackup.py:350]: Should we assert the `id` cannot be `0`?

File: test/suite/wtbackup.py [hunk 25]
```diff
@@ -343,12 +343,14 @@ def take_log_backup(self, bkup_c, backup_dir, orig_logs, log_cursor=None):
...
+            dest_id =  id
+        # Open the backup data source for incremental backup.    
+        config = 'incremental=(src_id="ID' +  str(src_id) + '",this_id="ID' + str(dest_id) + '"'
```
Comment (etienneptl) [inline test/suite/wtbackup.py:353]: We can use an `fstring` here:
```
config = f'incremental=(src_id="ID{src_id}",this_id="ID'{dest_id)}"'
```

File: test/suite/test_compact11.py [hunk 26]
```diff
@@ -163,24 +164,32 @@ def test_compact11(self):
...
+
+        bytes_recovered = 0
         while self.get_files_compacted(uris) < self.num_tables:
-            time.sleep(0.5)
+            new_bytes_recovered = self.get_bytes_recovered()
+            if new_bytes_recovered != bytes_recovered:
+                # Update the incremental backup ID from the parent class.
+                self.bkup_id += 1
+                shutil.copytree(self.home_tmp, self.backup_incr + str(self.bkup_id))
+                self.take_incr_backup(self.backup_incr + str(self.bkup_id), 0, self.bkup_id)
```
Comment (etienneptl) [inline test/suite/test_compact11.py:175]: Is it on purpose to always use `0` as the source? Is it because we change directory?

Comment (sueloverso) [reply test/suite/test_compact11.py:175]: I think it is because of my comment above to be able to compare during the loop and have a known base

**Review by sueloverso — APPROVED** [review]

**Review by etienneptl — APPROVED** [review]
LGTM, thank you very much for taking care of this!

File: test/suite/wtbackup.py (hunk 25 above)
Comment (jiechenbo) [reply test/suite/wtbackup.py]: [6933a2e](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/6933a2ea61497bafac665a9f342e788504d8b62f)

File: test/suite/wtbackup.py (hunk 24 above)
Comment (jiechenbo) [reply test/suite/wtbackup.py]: The assert is done at the start of the function.

File: test/suite/wtbackup.py (hunk 23 above)
Comment (jiechenbo) [reply test/suite/wtbackup.py:346]: [6933a2e](https://github.com/wiredtiger/wiredtiger/pull/9977/commits/6933a2ea61497bafac665a9f342e788504d8b62f)

**Comment by jiechenbo** [comment]
evergreen merge
